src/ast/ProgramDecl.ts

```diff
@@ -0,0 +1,38 @@
+import {ModuleDecl} from "./ModuleDecl";
+import {DirDecl} from "./DirDecl";
+import {Tokenizer} from "../util/Tokenizer";
+
+/**
+ * Represents a TypeScript project to be generated.
+ *
+ * Child of DirDecl since a TypeScript project is a directory with a name and modules.
+ *
+ * e.g. ‘project ’ (str | STRLIST) MODULE? (DIR|CLASS|INTERFACE)* nl
+ */
+export class ProgramDecl extends DirDecl {
```

Comment:
Not sure what the relationship between ProgramDecl and DirDecl is here. Maybe it would be clearer if we  had ProgramDecl have a Contents array (Dir | Interface | Class), and as part of ContentsDecl.parse() we can have some of the logic in the protected parseContents method? And Dir can extend Contents and call super.parse().

Reply:
So I used a Composite pattern from 210 to model the relationship between ```DirDecl``` and nested ```DirDecl``` or ```ClassDecl``` and ```InterfaceDecl``` inside it. And since the root ProgramDecl node we are treating as the root directory, it subclasses ```DirDecl``` nicely.

> Maybe it would be clearer if we had ProgramDecl have a Contents array (Dir | Interface | Class)

Not sure why we would add a Content array to ```ProgramDecl``` when it already implicitly has it by inheriting it from ```DirDecl```. 

> as part of ContentsDecl.parse() we can have some of the logic in the protected parseContents method
ContentsDecl.parse() doesn't exist, did you mean something else? What logic do you want to add to the protected parseContents() method?

I don't want to add ```DirDecl```/```ProgramDecl``` logic to ```Content``` because ```ClassDecl``` and ```InterfaceDecl``` inherit from ```Content``` as well and we don't want them to inherit that logic.

I hope this clears things up, let me know if there's anything else

Reply:
> Not sure why we would add a Content array to ProgramDecl when it already implicitly has it by inheriting it from DirDecl.
Sorry- that comment was unclear. I meant having a Content[] field with ProgramDecl not extending DirDecl.

> I don't want to add DirDecl/ProgramDecl logic to Content because ClassDecl and InterfaceDecl inherit from Content as well and we don't want them to inherit that logic.

I see- that makes sense. I was originally suggesting moving the shared logic for DirDecl and ProgramDecl to ContentsDecl, but I see your point about not wanting ClassDecl and InterfaceDecl to inherit it. 

Thanks for the clarification!